org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -1442,6 +1470,90 @@ private boolean isValid(@Nullable WorkspaceFolder wsFolder) {
 
 	}
 
+	/**
+	 * Resource listener that translates Eclipse resource change events into LSP file watch events
+	 * and dispatches them if the language server is still active
+	 */
+	private final class WatchedFilesListener implements IResourceChangeListener {
+
+		@Override
+		public void resourceChanged(final IResourceChangeEvent event) {
+			List<FileEvent> fileEvents = toFileEvents(event);
+			if (fileEvents.isEmpty()) {
+				return;
+			}
+			final LanguageServer currentServer = context.languageServer;
+			if (currentServer != null) {
+				currentServer.getWorkspaceService()
+						.didChangeWatchedFiles(new DidChangeWatchedFilesParams(fileEvents));
+			}
+		}
+
+		private List<FileEvent> toFileEvents(final IResourceChangeEvent event) {
+			if (event.getType() != IResourceChangeEvent.POST_CHANGE || event.getDelta() == null) {
+				return new ArrayList<>();
```

Comment:
```suggestion
				return List.of();
```
to avoid creating a new list?